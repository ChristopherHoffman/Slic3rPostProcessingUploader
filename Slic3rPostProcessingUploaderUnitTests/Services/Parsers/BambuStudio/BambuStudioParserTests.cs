using Slic3rPostProcessingUploader.Services.Parsers.AnycubicSlicerNext;
using Slic3rPostProcessingUploader.Services.Parsers.BambuStudio;
using Slic3rPostProcessingUploaderUnitTests.Services.Parsers.AnycubicSlicerNext;
using Snapshooter.MSTest;


namespace Slic3rPostProcessingUploaderUnitTests.Services.Parsers.BambuStudio
{
    [TestClass]
    public class BambuStudioParserTests
    {
        [TestMethod]
        public void ShouldReturnAnEmptySettingWhenGcodeIsEmpty()
        {
            var parser = new BambuStudioParser("");
            var result = parser.ParseGcode("");
            Assert.IsNotNull(result);

            Snapshot.Match(result);
        }

        [TestMethod]
        public void ShouldReturnExpectedValuesWhenGivenFullGcode()
        {
            var parser = new BambuStudioParser("");
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Snapshot.Match(result);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithNoReplacements()
        {
            string template = "Settings:";

            var parser = new BambuStudioParser(template);
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Assert.AreEqual("Settings:", result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithASingleReplacement()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                """;

            var parser = new BambuStudioParser(template);
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithMultipleReplacements()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                    First Layer Height: {{initial_layer_print_height}}
                    Wall Loops: {{wall_loops}}
                    Top Shell Layers: {{top_shell_layers}}
                    Bottom Shell Layers: {{bottom_shell_layers}}
                    Sparse Infill Density: {{sparse_infill_density}}
                """;

            var parser = new BambuStudioParser(template);
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                    First Layer Height: 0.28
                    Wall Loops: 3
                    Top Shell Layers: 3
                    Bottom Shell Layers: 3
                    Sparse Infill Density: 10%
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderFullTemplateWhenGivenAGcodeWithTwoFilaments()
        {
            var parser = new BambuStudioParser("");
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCubeTwoFilament);

            Snapshot.Match(result, matchOptions => matchOptions.HashField("settings.Snapshot"));
        }

    }
}
