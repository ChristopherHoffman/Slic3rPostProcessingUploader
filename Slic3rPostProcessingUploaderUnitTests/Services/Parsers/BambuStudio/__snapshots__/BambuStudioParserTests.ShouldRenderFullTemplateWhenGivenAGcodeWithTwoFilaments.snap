{
  "Slicer": "BambuStudioSlicer",
  "CuraVersion": "01.10.01.50",
  "PluginVersion": null,
  "settings": {
    "note": "Settings:\n\nQuality:\n  Layer Height:\n    Layer Height: 0.2\n    First Layer Height: 0.2\n  Line Width:\n    Default Line Width: 0.42\n  Seam:\n    Seam Position: aligned\n  Precision:\n    Precise Wall: \n  Ironing:\n    Ironing: no ironing\n  Wall Generator:\n    Generator: classic\n  Walls And Surfaces:\n    Only One Wall on Top Surfaces: \n\nStrength:\n  Walls:\n    Wall Loops: 2\n  Top/Bottom Shells:\n    Top Shell Layers: 5\n    Top Shell Thickness: 1\n    Bottom Shell Layers: 3\n    Bottom Shell Thickness: 0\n  Infill:\n    Sparse Infill Density: 15%\n    Sparse Infill Pattern: grid\n\nSpeed:\n  First Layer Speed:\n    First Layer: 50\n    Number of Slow Layers: \n  Other Layers Speed:\n    Outer Wall: 200\n    Inner Wall: 300\n    Sparse Infill: 270\n    Internal Solid Infill: 250\n    Top Surface: 200\n  Travel Speed:\n    Travel: 700\n\nSupport:\n  Support:\n    Enable Support: 0\n    Type: normal(auto)\n    Style: default\n    Threshold Angle: 30\n    On Build Plate Only: 0\n  Raft:\n    Raft Layers: 0\n\nMultimaterial:\n  Prime Tower:\n    Enable: 1\n    Width: 35\n    Brim Width: 3\n  Ooze Prevention:\n    Enable: 0\n  Flush Options:\n    Flush into Objects' Infill: 0\n    Flush into Objects' Support: 1\n  Advanced:\n    Use Beam Interlocking: \n\nOther:\n  Skirt:\n    Skirt Type: \n    Skirt Loops: 0\n    Skirt Min Extrusion Length: \n  Brim:\n    Brim Type: auto_brim\n    Brim Width: 5\n  Special Mode:\n    Slicing Mode: regular\n    Print Sequence: by layer\n    Spiral Vase: 0\n    Fuzzy Skin: none",
    "print_name": null,
    "time_stamp": null,
    "estimated_print_time_seconds": 15173,
    "material_used_mg": null,
    "Snapshot": "dCNOmK/nSY+12vHzasLXiswzlGT5UHA7jAGYkvmCuQs=",
    "file_name": null,
    "filamentUsage": [
      {
        "Id": null,
        "Filament": {
          "Id": "00000000-0000-0000-0000-000000000000",
          "DisplayName": "Other"
        },
        "AmountMg": null,
        "LengthInM": null,
        "VolumeMl": null,
        "EstimatedAmountMg": null,
        "EstimatedLengthInM": 9.035,
        "EstimatedVolumeMl": null,
        "Source": "Length",
        "EstimatedSource": "Length",
        "Notes": ""
      },
      {
        "Id": null,
        "Filament": {
          "Id": "00000000-0000-0000-0000-000000000000",
          "DisplayName": "Other"
        },
        "AmountMg": null,
        "LengthInM": null,
        "VolumeMl": null,
        "EstimatedAmountMg": null,
        "EstimatedLengthInM": 11.468,
        "EstimatedVolumeMl": null,
        "Source": "Length",
        "EstimatedSource": "Length",
        "Notes": ""
      }
    ]
  }
}
